core/src/apps/bitcoin/ownership.py

```diff
@@ -0,0 +1,132 @@
+from trezor import utils, wire
+from trezor.crypto import bip32, hashlib, hmac
+
+from apps.bitcoin.multisig import multisig_pubkey_index
+from apps.bitcoin.writers import write_bytes_prefixed
+from apps.common import seed
+from apps.common.readers import BytearrayReader, read_bitcoin_varint
+from apps.common.writers import (
+    empty_bytearray,
+    write_bitcoin_varint,
+    write_bytes_fixed,
+    write_uint8,
+)
+
+from . import common, scripts
+from .readers import read_bytes_prefixed
+from .verification import SignatureVerifier
+
+if False:
+    from typing import List, Optional, Tuple
+    from trezor.messages.MultisigRedeemScriptType import MultisigRedeemScriptType
+    from trezor.messages.TxInputType import EnumTypeInputScriptType
+    from apps.common.coininfo import CoinInfo
+
+# This module implements the SLIP-0019 proof of ownership format.
+
+_VERSION_MAGIC = b"SL\x00\x19"
+_FLAG_USER_CONFIRMED = 0x01
+_OWNERSHIP_ID_LEN = 32
+_OWNERSHIP_ID_KEY_PATH = [b"SLIP-0019", b"Ownership identification key"]
+
+
+def generate_proof(
+    node: bip32.HDNode,
+    script_type: EnumTypeInputScriptType,
+    multisig: MultisigRedeemScriptType,
+    coin: CoinInfo,
+    user_confirmed: bool,
+    ownership_ids: List[bytes],
+    script_pubkey: bytes,
+    commitment_data: Optional[bytes],
+) -> Tuple[bytes, bytes]:
+    flags = 0
+    if user_confirmed:
+        flags |= _FLAG_USER_CONFIRMED
+
+    proof = empty_bytearray(4 + 1 + 1 + len(ownership_ids) * _OWNERSHIP_ID_LEN)
+
+    write_bytes_fixed(proof, _VERSION_MAGIC, 4)
+    write_uint8(proof, flags)
+    write_bitcoin_varint(proof, len(ownership_ids))
+    for ownership_id in ownership_ids:
+        write_bytes_fixed(proof, ownership_id, _OWNERSHIP_ID_LEN)
+
+    sighash = hashlib.sha256(proof)
+    sighash.update(script_pubkey)
+    if commitment_data:
+        sighash.update(commitment_data)
+    signature = common.ecdsa_sign(node, sighash.digest())
+    public_key = node.public_key()
+
+    script_sig = scripts.input_derive_script(
+        script_type, multisig, coin, common.SIGHASH_ALL, public_key, signature
+    )
+    if script_type in common.SEGWIT_INPUT_SCRIPT_TYPES:
+        if multisig:
+            # find the place of our signature based on the public key
+            signature_index = multisig_pubkey_index(multisig, public_key)
+            witness = scripts.witness_p2wsh(
+                multisig, signature, signature_index, common.SIGHASH_ALL
+            )
+        else:
+            witness = scripts.witness_p2wpkh(signature, public_key, common.SIGHASH_ALL)
+    else:
+        # Zero entries in witness stack.
+        witness = b"\x00"
```

Comment:
What about creating a new method that BIP322 signs messages.

Reply:
f33772013b071791e4002c5f6f3088c37e90af13, f2ea0e42731c2edc54f93bae916fa53d0069a360